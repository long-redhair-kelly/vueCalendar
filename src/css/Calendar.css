#main {
  font-size: 14px;
  line-height: 20px;
  table-layout: fixed;
  width: 100%;
  margin-bottom: 1rem;
  color: silver;
  border-bottom: 1px solid #ddd;
  border-collapse: collapse;
}
#main th {
  padding: 0;
  text-align: center;
  vertical-align: middle;
  font-weight: normal;
  color: #333;
}
#main td {
  padding: 8px;
  text-align: center;
  vertical-align: middle;
  border-top: 1px solid #ddd;
}
thead {
  background-color: #f4c777;
}
.today {
  background-color: #fcf8e3;
}
