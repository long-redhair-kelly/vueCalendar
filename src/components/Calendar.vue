<script setup lang="ts">
import "../css/Calendar.css";
import { ref, computed, onMounted } from "vue";
import { YearMonth } from "../../src/interfaces/YearMonth";

const weekDays = ["日", "月", "火", "水", "木", "金", "土"];
const currentDate = new Date();
let yearMonth = ref<YearMonth>({
  year: currentDate.getFullYear(),
  month: currentDate.getMonth() + 1,
  today: "",
});

/**
 * カレンダー本体
 */
const computedCalendar = computed(() => {
  let calendar: Array<Array<string>> = [];
  let firstWeekDay = new Date(
    yearMonth.value.year,
    yearMonth.value.month - 1,
    1
  ).getDay();
  let lastDay = new Date(
    yearMonth.value.year,
    yearMonth.value.month,
    0
  ).getDate();
  let dayNumber = 1;

  while (dayNumber <= lastDay) {
    let weekData: Array<string> = [];
    for (let i = 0; i <= 6; i++) {
      if (calendar.length == 0 && i < firstWeekDay) {
        weekData[i] = "";
      } else if (lastDay < dayNumber) {
        weekData[i] = "";
      } else {
        weekData[i] = dayNumber.toString();
        dayNumber++;
      }
    }
    calendar.push(weekData);
  }
  return calendar;
});

/**
 * 現在年月
 */
const currentYearMonth = () => {
  yearMonth.value.year = currentDate.getFullYear();
  yearMonth.value.month = currentDate.getMonth() + 1;
  return yearMonth;
};

/**
 * 前年
 */
const prevYear = () => {
  yearMonth.value.year--;
  return yearMonth;
};

/**
 * 前月
 */
const prevMonth = () => {
  yearMonth.value.month--;
  if (yearMonth.value.month < 1) {
    yearMonth.value.year--;
    yearMonth.value.month = 12;
  }
  return yearMonth;
};

/**
 * 来月
 */
const nextMonth = () => {
  yearMonth.value.month++;
  if (yearMonth.value.month > 12) {
    yearMonth.value.year++;
    yearMonth.value.month = 1;
  }
  return yearMonth;
};

/**
 * 来年
 */
const nextYear = () => {
  yearMonth.value.year++;
  return yearMonth;
};

/**
 * 指定した日が今日か判定
 * @param day 日付
 * @return true: 今日、false: 今日以外
 */
const isToday = (day: string) => {
  let date = yearMonth.value.year + "-" + yearMonth.value.month + "-" + day;
  if (yearMonth.value.today == date) {
    return true;
  }
  return false;
};

onMounted(() => {
  let date = new Date();
  yearMonth.value.year = date.getFullYear();
  yearMonth.value.month = date.getMonth() + 1;
  let actualDay = date.getDate();
  yearMonth.value.today =
    yearMonth.value.year + "-" + yearMonth.value.month + "-" + actualDay;
});
</script>

<template>
  <div>
    <div id="header">
      <span class="header-arrow" @click="prevYear">◀</span>
      <span class="header-arrow" @click="prevMonth">＜</span>
      <span class="selected-month" @click="currentYearMonth"
        >{{ yearMonth.year }}年{{ yearMonth.month }}月</span
      >
      <span class="header-arrow" @click="nextMonth">＞</span>
      <span class="header-arrow" @click="nextYear">▶</span>
    </div>
    <table id="main">
      <thead>
        <th v-for="(weekDay, weekDayIndex) in weekDays" :key="weekDayIndex">
          {{ weekDay }}
        </th>
      </thead>
      <tbody>
        <tr
          v-for="(rowNumber, rowNumberIndex) in computedCalendar"
          :key="rowNumberIndex"
        >
          <td
            v-for="(columnNumber, columnNumberIndex) in rowNumber"
            :key="columnNumberIndex"
            :class="{ today: isToday(columnNumber) }"
          >
            <span v-if="isToday(columnNumber)">today</span>
            <span v-else>{{ columnNumber }}</span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</template>
