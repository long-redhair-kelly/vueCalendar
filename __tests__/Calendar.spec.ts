import { render } from "@testing-library/vue";
import Calendar from "../src/components/Calendar";

describe("Calendar.vue", () => {
  it("Calendar スナップショット(currentDateだけがNGの場合「yarn test:update」を実施)", () => {
    const wrapper = render(Calendar);
    expect(wrapper).toMatchSnapshot();
  });
});
