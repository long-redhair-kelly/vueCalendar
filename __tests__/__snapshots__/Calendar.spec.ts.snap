// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar.vue Calendar スナップショット(currentDateだけがNGの場合「yarn test:update」を実施) 1`] = `
{
  "baseElement": <body>
    <div>
      <div>
        <div
          id="header"
        >
          <span
            class="header-arrow"
          >
            ◀
          </span>
          <span
            class="header-arrow"
          >
            ＜
          </span>
          <span
            class="selected-month"
          >
            2023年4月
          </span>
          <span
            class="header-arrow"
          >
            ＞
          </span>
          <span
            class="header-arrow"
          >
            ▶
          </span>
        </div>
        <table
          id="main"
        >
          <thead>
            
            <th>
              日
            </th>
            <th>
              月
            </th>
            <th>
              火
            </th>
            <th>
              水
            </th>
            <th>
              木
            </th>
            <th>
              金
            </th>
            <th>
              土
            </th>
            
          </thead>
          <tbody>
            
            <tr>
              
              <td
                class=""
              >
                <span />
              </td>
              <td
                class=""
              >
                <span />
              </td>
              <td
                class=""
              >
                <span />
              </td>
              <td
                class=""
              >
                <span />
              </td>
              <td
                class=""
              >
                <span />
              </td>
              <td
                class=""
              >
                <span />
              </td>
              <td
                class=""
              >
                <span>
                  1
                </span>
              </td>
              
            </tr>
            <tr>
              
              <td
                class=""
              >
                <span>
                  2
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  3
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  4
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  5
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  6
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  7
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  8
                </span>
              </td>
              
            </tr>
            <tr>
              
              <td
                class=""
              >
                <span>
                  9
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  10
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  11
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  12
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  13
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  14
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  15
                </span>
              </td>
              
            </tr>
            <tr>
              
              <td
                class=""
              >
                <span>
                  16
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  17
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  18
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  19
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  20
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  21
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  22
                </span>
              </td>
              
            </tr>
            <tr>
              
              <td
                class=""
              >
                <span>
                  23
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  24
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  25
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  26
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  27
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  28
                </span>
              </td>
              <td
                class=""
              >
                <span>
                  29
                </span>
              </td>
              
            </tr>
            <tr>
              
              <td
                class=""
              >
                <span>
                  30
                </span>
              </td>
              <td
                class=""
              >
                <span />
              </td>
              <td
                class=""
              >
                <span />
              </td>
              <td
                class=""
              >
                <span />
              </td>
              <td
                class=""
              >
                <span />
              </td>
              <td
                class=""
              >
                <span />
              </td>
              <td
                class=""
              >
                <span />
              </td>
              
            </tr>
            
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        id="header"
      >
        <span
          class="header-arrow"
        >
          ◀
        </span>
        <span
          class="header-arrow"
        >
          ＜
        </span>
        <span
          class="selected-month"
        >
          2023年4月
        </span>
        <span
          class="header-arrow"
        >
          ＞
        </span>
        <span
          class="header-arrow"
        >
          ▶
        </span>
      </div>
      <table
        id="main"
      >
        <thead>
          
          <th>
            日
          </th>
          <th>
            月
          </th>
          <th>
            火
          </th>
          <th>
            水
          </th>
          <th>
            木
          </th>
          <th>
            金
          </th>
          <th>
            土
          </th>
          
        </thead>
        <tbody>
          
          <tr>
            
            <td
              class=""
            >
              <span />
            </td>
            <td
              class=""
            >
              <span />
            </td>
            <td
              class=""
            >
              <span />
            </td>
            <td
              class=""
            >
              <span />
            </td>
            <td
              class=""
            >
              <span />
            </td>
            <td
              class=""
            >
              <span />
            </td>
            <td
              class=""
            >
              <span>
                1
              </span>
            </td>
            
          </tr>
          <tr>
            
            <td
              class=""
            >
              <span>
                2
              </span>
            </td>
            <td
              class=""
            >
              <span>
                3
              </span>
            </td>
            <td
              class=""
            >
              <span>
                4
              </span>
            </td>
            <td
              class=""
            >
              <span>
                5
              </span>
            </td>
            <td
              class=""
            >
              <span>
                6
              </span>
            </td>
            <td
              class=""
            >
              <span>
                7
              </span>
            </td>
            <td
              class=""
            >
              <span>
                8
              </span>
            </td>
            
          </tr>
          <tr>
            
            <td
              class=""
            >
              <span>
                9
              </span>
            </td>
            <td
              class=""
            >
              <span>
                10
              </span>
            </td>
            <td
              class=""
            >
              <span>
                11
              </span>
            </td>
            <td
              class=""
            >
              <span>
                12
              </span>
            </td>
            <td
              class=""
            >
              <span>
                13
              </span>
            </td>
            <td
              class=""
            >
              <span>
                14
              </span>
            </td>
            <td
              class=""
            >
              <span>
                15
              </span>
            </td>
            
          </tr>
          <tr>
            
            <td
              class=""
            >
              <span>
                16
              </span>
            </td>
            <td
              class=""
            >
              <span>
                17
              </span>
            </td>
            <td
              class=""
            >
              <span>
                18
              </span>
            </td>
            <td
              class=""
            >
              <span>
                19
              </span>
            </td>
            <td
              class=""
            >
              <span>
                20
              </span>
            </td>
            <td
              class=""
            >
              <span>
                21
              </span>
            </td>
            <td
              class=""
            >
              <span>
                22
              </span>
            </td>
            
          </tr>
          <tr>
            
            <td
              class=""
            >
              <span>
                23
              </span>
            </td>
            <td
              class=""
            >
              <span>
                24
              </span>
            </td>
            <td
              class=""
            >
              <span>
                25
              </span>
            </td>
            <td
              class=""
            >
              <span>
                26
              </span>
            </td>
            <td
              class=""
            >
              <span>
                27
              </span>
            </td>
            <td
              class=""
            >
              <span>
                28
              </span>
            </td>
            <td
              class=""
            >
              <span>
                29
              </span>
            </td>
            
          </tr>
          <tr>
            
            <td
              class=""
            >
              <span>
                30
              </span>
            </td>
            <td
              class=""
            >
              <span />
            </td>
            <td
              class=""
            >
              <span />
            </td>
            <td
              class=""
            >
              <span />
            </td>
            <td
              class=""
            >
              <span />
            </td>
            <td
              class=""
            >
              <span />
            </td>
            <td
              class=""
            >
              <span />
            </td>
            
          </tr>
          
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "emitted": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "html": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
